fe/src/main/java/org/apache/doris/load/routineload/RoutineLoadTaskScheduler.java

```diff
@@ -28,52 +28,74 @@
  * Step1: get total idle task num of backends.
+ *   Step1.1: if total idle task num == 0, exit this round and switch to the next round immediately
  * Step2: equally divide to be
+ *   Step2.1: if there is no task in queue, waiting task until an element becomes available.
+ *   Step2.2: divide task to be
  * Step3: submit tasks to be
  */
-// TODO(ml): change interval ms in constructor
 public class RoutineLoadTaskScheduler extends Daemon {
 
     private static final Logger LOG = LogManager.getLogger(RoutineLoadTaskScheduler.class);
 
-    private RoutineLoadManager routineLoadManager = Catalog.getInstance().getRoutineLoadManager();
+    private RoutineLoadManager routineLoadManager;
+    private LinkedBlockingQueue<RoutineLoadTaskInfo> needSchedulerTasksQueue;
+
+    public RoutineLoadTaskScheduler() {
+        super("routine load task", 0);
+        routineLoadManager = Catalog.getInstance().getRoutineLoadManager();
+        needSchedulerTasksQueue = (LinkedBlockingQueue) routineLoadManager.getNeedSchedulerTasksQueue();
+    }
 
     @Override
     protected void runOneCycle() {
         try {
             process();
         } catch (Throwable e) {
             LOG.warn("Failed to process one round of RoutineLoadTaskScheduler with error message {}",
-                    e.getMessage(), e);
+                     e.getMessage(), e);
         }
     }
 
     private void process() throws LoadException {
         // update current beIdMaps for tasks
         routineLoadManager.updateBeIdTaskMaps();
 
-        // get idle be task num
+        LOG.info("There are {} need scheduler task in queue when {}",
+                 needSchedulerTasksQueue.size(), System.currentTimeMillis());
+        AgentBatchTask batchTask = new AgentBatchTask();
+        int sizeOfTasksQueue = needSchedulerTasksQueue.size();
         int clusterIdleSlotNum = routineLoadManager.getClusterIdleSlotNum();
+        int needScheduledTaskNum = sizeOfTasksQueue < clusterIdleSlotNum ? sizeOfTasksQueue : clusterIdleSlotNum;
         int scheduledTaskNum = 0;
-        Queue<RoutineLoadTaskInfo> needSchedulerTasksQueue = routineLoadManager.getNeedSchedulerTasksQueue();
-        AgentBatchTask batchTask = new AgentBatchTask();
-
+        // get idle be task num
         // allocate task to be
-        while (clusterIdleSlotNum > 0) {
-            if (needSchedulerTasksQueue.peek() != null) {
-                RoutineLoadTaskInfo routineLoadTaskInfo = needSchedulerTasksQueue.poll();
+        while (needScheduledTaskNum > 0) {
+            RoutineLoadTaskInfo routineLoadTaskInfo = null;
+            try {
+                routineLoadTaskInfo = needSchedulerTasksQueue.take();
+            } catch (InterruptedException e) {
+                LOG.warn("Taking routine load task from queue has been interrupted with error msg {}",
+                         e.getMessage());
+                return;
+            }
+
+            if (clusterIdleSlotNum > 0) {
```

Comment:
You don't need to check if clusterIdleSlotNum > 0 here. Because if clusterIdleSlotNum <= 0, you will not enter this while loop at all. And you already make sure that clusterIdleSlotNum is always larger than needScheduledTaskNum.